src/main/java/com/atlassian/bitbucket/jenkins/internal/applink/oauth/serviceprovider/token/PersistentServiceProviderTokenStore.java

```diff
@@ -0,0 +1,422 @@
+    }
+
+    private static final class ServiceProviderTokenConverter implements Converter {
+
+        private static final String ACCESS_TOKEN = "access-token";
+        private static final String TOKEN_VALUE = "token-value";
+        private static final String TOKEN_SECRET = "token-secret";
+        private static final String CONSUMER_KEY = "consumer-key";
+        private static final String PROPERTIES = "properties";
+        private static final String AUTHORIZATION = "authorization";
+        private static final String USER = "user";
+        private static final String VERIFIER = "verifier";
+        private static final String TOKEN_CREATION_TIME = "creation-time";
+        private static final String TOKEN_TIME_TO_LIVE = "time-to-live";
+        private static final String CALLBACK = "callback";
+        private static final String SESSION = "session";
+        private static final String SESSION_HANDLE = "handle";
+        private static final String SESSION_CREATION_TIME = "creation-time";
+        private static final String SESSION_LAST_RENEWAL_TIME = "last-renewal-time";
+        private static final String SESSION_TIME_TO_LIVE = "time-to-live";
+
+        private final ConsumerStore consumerStore;
+
+        private ServiceProviderTokenConverter(ConsumerStore consumerStore) {
+            this.consumerStore = consumerStore;
+        }
+
+        private static void addProperties(HierarchicalStreamWriter writer, Map<String, String> properties) {
+            if (properties != null) {
+                writer.startNode(PROPERTIES);
+                properties.forEach((key, value) -> addNode(writer, key, value));
+                writer.endNode();
+            }
+        }
+
+        private static void addSession(HierarchicalStreamWriter writer, ServiceProviderToken.Session session) {
+            if (session != null) {
+                writer.startNode(SESSION);
+                addNode(writer, SESSION_HANDLE, encrypt(session.getHandle()));
+                addNode(writer, SESSION_CREATION_TIME, session.getCreationTime());
+                addNode(writer, SESSION_LAST_RENEWAL_TIME, session.getLastRenewalTime());
+                addNode(writer, SESSION_TIME_TO_LIVE, session.getTimeToLive());
+                writer.endNode();
+            }
+        }
+
+        private static void addNode(HierarchicalStreamWriter writer, String name, Object value) {
+            writer.startNode(name);
+            writer.setValue(Objects.toString(value));
+            writer.endNode();
+        }
+
+        private static Map<String, String> unmarshalProperties(HierarchicalStreamReader reader) {
+            Map<String, String> properties = new HashMap<>();
+            while (reader.hasMoreChildren()) {
+                reader.moveDown();
+                properties.put(reader.getNodeName(), reader.getValue());
+                reader.moveUp();
+            }
+            return properties;
+        }
+
```

Comment:
Do we usually use these markers for private methods?

Reply:
That's a good question. I believe we use them everywhere (and can find examples in the code base), but might be worth double-checking

Reply:
You're right, we have a few on private methods. There's also a mix of `@Nullable` and `@CheckForNull` which seem to do the same thing? I'd probably go for `@Nullable` since it's a standard Java annotation, unless someone can tell me there is a specific benefit to the other one in Jenkins.

Reply:
From my understanding, `@CheckForNull` _requires_ the consumer to make a null check (i.e. a null value in that situation is not safe), while `@Nullable` seems to simply say a null input/output is possible- but that doesn't seem to be how we use it in the code. @goravjoshi ?

Reply:
I _think_ in this case (and definitely check with someone else), the value can't really be used without attempting to dereference it so it should be `@CheckForNull`... but then isn't that true for _all_ nullable methods? :confused: